examples/iterators_and_for_range/example.cpp

```diff
@@ -0,0 +1,113 @@
+  std::cout << "Value of it_rbegin: "<< *it_rbegin << std::endl;
+  // Iterator pointing one position after the last element.
+  std::vector<int>::reverse_iterator it_rend = vector_int.rend();
+  // Same way, "Segmentation Fault" could be appear here:
+  // std::cout << "Value of it_rend: "<< *it_rend << std::endl;
+
+  // There are CONST reverse iterator too.
+  std::vector<int>::const_reverse_iterator it_crbegin = vector_int.crbegin();
+  std::vector<int>::const_reverse_iterator it_crend = vector_int.crend();
+  // Note: `Auto` keyword is your friend, there is no need to be extra verbose.
+
+
+  ////////////////////////////////
+  ////    Applying knowledge  ////
+  ////////////////////////////////
+
+  std::cout << "----- Modifying vector_int ------" << std::endl;
+  for (auto it = vector_int.begin(); it != vector_int.end(); it++){
+    *it = *it * 2;
+  }
+
+  std::cout << "----- Reading vector_int ------" << std::endl;
+  // Given that we just want to read the values, we could use cbegin and cend.
+  for (auto it = vector_int.cbegin(); it != vector_int.cend(); it++){
+    std::cout << *it << std::endl;
+  }
+
+  std::cout << "----- Reading vector_int backwars ------" << std::endl;
+  // Given that we just want to read the values, we could use crbegin and crend.
+  for (auto it = vector_int.crbegin(); it != vector_int.crend(); it++){
+    std::cout << *it << std::endl;
+  }
+
+  std::cout << "----- Range based for loop ------" << std::endl;
+  // Range based for loop could be used to avoid extra verbose expressions.
+  // In the background, iterators are being used to iterate this loop.
+  std::cout << " - `value` contains a copy of the i-th value of `vector_int`"<< std::endl;
+  for(auto value : vector_int){
+    // Values within `vector_int` won't be changed!.
+    value = 0;
+  }
+  // Check those same values as before.
+  for(auto value : vector_int){
+    std::cout << value << std::endl;
+  }
+  std::cout << " - `value` contains a reference of the i-th value of `vector_int`"<< std::endl;
+  // But if you want to modify them you can use reference instead.
+  for(auto& value : vector_int){
+    // Values within `vector_int` will be splitted!.
+    value = value/2;
+  }
+  for(auto& value : vector_int){
+    std::cout << value << std::endl;
+  }
+  // If you just want to read the values, then a constant reference is the appropiate, in order to avoid performance drop.
+  // `const` is added in order to avoid modifying the values by mistake.
+  std::cout << " - `value` contains a constant reference of the i-th value of `vector_int`"<< std::endl;
+  for(const auto& value : vector_int){
+    std::cout << value << std::endl;
+  }
+
+  std::cout << "----- Iterate old c-style arrays ------" << std::endl;
+  // Can you iterate old c-style arrays this way? Sure!
+  const std::string old_vector[4]{"uno", "dos", "tres", "c++"};
+  for (const auto& value : old_vector){
+    std::cout << value << std::endl;
+  }
+
```

Comment:
Validity of iterators might be worth mentioning, with some examples of operations that invalidate them.

Reply:
Let me check if was clear enough.
Done.